staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/fields_proto.go

```diff
@@ -0,0 +1,88 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"encoding/json"
+)
+
+// Fields is declared in types.go
+
+// ProtoFields is a struct that is equivalent to Fields, but intended for
+// protobuf marshalling/unmarshalling. It is generated into a serialization
+// that matches Fields. Do not use in Go structs.
+type ProtoFields struct {
+	// Map is the representation used in the alpha version of this API
+	Map map[string]Fields `json:"-" protobuf:"bytes,1,rep,name=map"`
+
+	// Raw is the underlying serialization of this object.
+	Raw []byte `json:"-" protobuf:"bytes,2,opt,name=raw"`
+}
+
+// ProtoFields returns the Fields as a new ProtoFields value.
+func (m *Fields) ProtoFields() *ProtoFields {
+	if m == nil {
+		return &ProtoFields{}
+	}
+	return &ProtoFields{
+		Raw: m.Raw,
+	}
+}
+
+// Size implements the protobuf marshalling interface.
+func (m *Fields) Size() (n int) {
+	return m.ProtoFields().Size()
+}
+
+// Unmarshal implements the protobuf marshalling interface.
+func (m *Fields) Unmarshal(data []byte) error {
+	if len(data) == 0 {
+		return nil
+	}
+	p := ProtoFields{}
+	if err := p.Unmarshal(data); err != nil {
+		return err
+	}
+	if len(p.Map) == 0 {
+		return json.Unmarshal(p.Raw, &m)
```

Comment:
I'm having a hard time following this... some comments might help.

if I'm reading this correctly, we expect this to be the normal path (since we should not encounter persisted alpha data)

does that mean our preferred proto format is a byte array of json? I didn't realize that's where we ended up.

Reply:
this also potentially forces another allocation of len(p.Raw), since Fields#UnmarshalJSON just appends non-null bytes to the Raw field (with `f.Raw = append(f.Raw[0:0], b...)`).

Would this work better?

```go
m.Raw = nil
if !bytes.Equal(p.Raw, []byte("null")) {
  m.Raw = p.Raw
}
```

Reply:
And a follow up for this too